SIGSEGV when running builds

### Contributing guidelines and issue reporting guide

- [x] I've read the [contributing guidelines](https://github.com/moby/buildkit/blob/master/.github/CONTRIBUTING.md) and wholeheartedly agree. I've also read the [issue reporting guide](https://github.com/moby/buildkit/blob/master/.github/issue_reporting_guide.md).

### Well-formed report checklist

- [x] I have found a bug that the documentation does not mention anything about my problem
- [x] I have found a bug that there are no open or closed issues that are related to my problem
- [x] I have provided version/information about my environment and done my best to provide a reproducer

### Description of bug

Hello :-)

We're running builds and sometimes our server is returning `401` status codes to `ADD`s.

This crashes the buildkit daemon with a SIGSEGV, e.g.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1bac412]
```

It doesn't happen 100% of the time, but it happens pretty often...

We've reproduced this on at least two boxes, and have created a small reproduction: https://gist.github.com/bglimepoint/a2dec1a5bc30646b26b9846e3b350ddb

This is happening on Fedora 43 with `Docker version 29.3.1`, and Ubuntu 26.0, with `Docker version 29.1.3`

This happens against the latest stable image as well. Our code doesn't use the `--checksum` arg, but it was needed for the reproduction...

Thanks for your time :-)

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -116,6 +116,7 @@ var allTests = []func(t *testing.T, sb integration.Sandbox){
 	testCallDiskUsage,
 	testBuildMultiMount,
 	testBuildHTTPSource,
+	testBuildHTTPSourceUnauthorizedChecksumRace,
 	testBuildHTTPSourceEtagScope,
 	testBuildHTTPSourceAuthHeaderSecret,
 	testBuildHTTPSourceHostTokenSecret,
@@ -3783,6 +3784,50 @@ func testBuildHTTPSource(t *testing.T, sb integration.Sandbox) {
 	checkAllReleasable(t, c, sb, true)
 
 	// TODO: check that second request was marked as cached
+}
+
+func testBuildHTTPSourceUnauthorizedChecksumRace(t *testing.T, sb integration.Sandbox) {
+	integration.SkipOnPlatform(t, "windows")
+
+	c, err := New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	var requests atomic.Int32
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if requests.Add(1) == 1 {
+			time.Sleep(200 * time.Millisecond)
+		}
+		w.WriteHeader(http.StatusUnauthorized)
+	}))
+	defer server.Close()
+
+	url := server.URL + "/test.txt"
+	lower := llb.Scratch()
+	// Reproduce the race between same-URL HTTP sources with and without
+	// checksum sharing the resolver cache while the server returns 401.
+	withoutChecksum := lower.File(llb.Copy(llb.HTTP(url), "test.txt", "/a.txt"))
+	withChecksum := lower.File(llb.Copy(llb.HTTP(url, llb.Checksum(digest.FromBytes(nil))), "test.txt", "/b.txt"))
+	st := llb.Merge([]llb.State{
+		llb.Diff(lower, withoutChecksum),
+		llb.Diff(lower, withChecksum),
+	})
+
+	def, err := st.Marshal(sb.Context())
+	require.NoError(t, err)
+
+	for range 10 {
+		_, err := c.Solve(sb.Context(), def, SolveOpt{
+			Exports: []ExportEntry{
+				{
+					Type:      ExporterLocal,
+					OutputDir: t.TempDir(),
+				},
+			},
+		}, nil)
+		require.Error(t, err)
+		require.ErrorContains(t, err, "invalid response status 401")
+	}
 }
 
 // docker/buildx#2803

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -3817,6 +3817,7 @@ func testBuildHTTPSourceUnauthorizedChecksumRace(t *testing.T, sb integration.Sa
 	require.NoError(t, err)
 
 	for range 10 {
+		requests.Store(0)
 		_, err := c.Solve(sb.Context(), def, SolveOpt{
 			Exports: []ExportEntry{
 				{

--- a/solver/resolvercache.go
+++ b/solver/resolvercache.go
@@ -127,7 +127,9 @@ func (c *combinedCache) Lock(key any) (values []any, release func(any) error, er
 
 			mu.Lock()
 			valuesAll = append(valuesAll, vals...)
-			releasers = append(releasers, rel)
+			if rel != nil {
+				releasers = append(releasers, rel)
+			}
 			mu.Unlock()
 		}(rc)
 	}

--- a/solver/resolvercache.go
+++ b/solver/resolvercache.go
@@ -61,7 +61,7 @@ func (r *resolverCache) Lock(key any) (values []any, release func(any) error, er
 	defer r.mu.Unlock()
 	e2, ok := r.locks[key]
 	if !ok {
-		return nil, nil, nil // key deleted
+		return nil, func(any) error { return nil }, nil // key deleted
 	}
 	values = slices.Clone(e2.values)
 	if e2.locked {

--- a/source/http/source.go
+++ b/source/http/source.go
@@ -868,9 +868,9 @@ func (hs *httpSourceHandler) Snapshot(ctx context.Context, jobCtx solver.JobCont
 	}
 
 	if refID != "" {
-		ref, err := hs.cache.Get(ctx, hs.resolved.refID, nil)
+		ref, err := hs.cache.Get(ctx, refID, nil)
 		if err != nil {
-			bklog.G(ctx).WithError(err).Warnf("failed to get HTTP snapshot for ref %s (%s)", hs.resolved.refID, hs.src.URL)
+			bklog.G(ctx).WithError(err).Warnf("failed to get HTTP snapshot for ref %s (%s)", refID, hs.src.URL)
 		} else {
 			return ref, nil
 		}

--- a/source/http/source.go
+++ b/source/http/source.go
@@ -866,8 +866,8 @@ func (hs *httpSourceHandler) Snapshot(ctx context.Context, jobCtx solver.JobCont
 			}
 			for _, v := range vals {
 				v2, ok := v.(*metadataWithRef)
-				if !ok {
-					return nil, errors.Errorf("invalid HTTP resolver cache value: %T", vals[0])
+				if !ok || v2 == nil {
+					return nil, errors.Errorf("invalid HTTP resolver cache value: %T", v)
 				}
 				if hs.src.Checksum != "" && v2.Digest != hs.src.Checksum {
 					continue

--- a/source/http/source.go
+++ b/source/http/source.go
@@ -323,9 +323,9 @@ func (hs *httpSourceHandler) resolveMetadataRef(ctx context.Context, jobCtx solv
 			}
 			saveResolved := true
 			defer func() {
-				ret := hs.resolved
-				if retErr != nil || !saveResolved {
-					ret = nil
+				var ret any
+				if retErr == nil && saveResolved && hs.resolved != nil {
+					ret = hs.resolved
 				}
 				if err := release(ret); err != nil {
 					bklog.G(ctx).WithError(err).Warn("failed to release resolver cache lock")

--- a/source/http/source.go
+++ b/source/http/source.go
@@ -860,26 +860,9 @@ func (hs *httpSourceHandler) Snapshot(ctx context.Context, jobCtx solver.JobCont
 			if err != nil {
 				return nil, err
 			}
-			vals, release, err := rc.Lock(uh)
+			refID, err = hs.snapshotRefIDFromResolverCache(rc, uh)
 			if err != nil {
 				return nil, err
-			}
-			for _, v := range vals {
-				v2, ok := v.(*metadataWithRef)
-				if !ok || v2 == nil {
-					return nil, errors.Errorf("invalid HTTP resolver cache value: %T", v)
-				}
-				if hs.src.Checksum != "" && v2.Digest != hs.src.Checksum {
-					continue
-				}
-				if v2.refID != "" {
-					hs.resolved = v2
-					refID = v2.refID
-				}
-			}
-			release(nil)
-			if hs.src.Checksum != "" && len(vals) > 0 && refID == "" {
-				return nil, errors.Errorf("digest mismatch for %s: %s (expected: %s)", hs.src.URL, vals[0], hs.src.Checksum)
 			}
 		}
 	}
@@ -923,6 +906,33 @@ func (hs *httpSourceHandler) Snapshot(ctx context.Context, jobCtx solver.JobCont
 	}
 
 	return ref, nil
+}
+
+func (hs *httpSourceHandler) snapshotRefIDFromResolverCache(rc solver.ResolverCache, uh digest.Digest) (string, error) {
+	vals, release, err := rc.Lock(uh)
+	if err != nil {
+		return "", err
+	}
+	defer release(nil)
+
+	refID := ""
+	for _, v := range vals {
+		v2, ok := v.(*metadataWithRef)
+		if !ok || v2 == nil {
+			return "", errors.Errorf("invalid HTTP resolver cache value: %T", v)
+		}
+		if hs.src.Checksum != "" && v2.Digest != hs.src.Checksum {
+			continue
+		}
+		if v2.refID != "" {
+			hs.resolved = v2
+			refID = v2.refID
+		}
+	}
+	if hs.src.Checksum != "" && len(vals) > 0 && refID == "" {
+		return "", errors.Errorf("digest mismatch for %s: %s (expected: %s)", hs.src.URL, vals[0], hs.src.Checksum)
+	}
+	return refID, nil
 }
 
 func (hs *httpSourceHandler) newHTTPRequest(ctx context.Context, g session.Group) (*http.Request, error) {

